src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -56,16 +56,8 @@ public JsonWebToken(string jwtEncodedString)
             if (string.IsNullOrEmpty(jwtEncodedString))
                 throw new ArgumentNullException(nameof(jwtEncodedString));
 
-            int count = 1;
-            int next = -1;
-            while ((next = jwtEncodedString.IndexOf('.', next + 1)) != -1)
-            {
-                count++;
-                if (count >= JwtConstants.JwsSegmentCount)
-                    break;
-            }
-
-            // JWS or JWE
+            // Number of segments is always one more than the number of '.'s.
+            var count = jwtEncodedString.Count(c => c == '.') + 1;
```

Comment:
What happens with a large number of dots, say 100,000?
This was the reason to avoid split.
You can change the condition to count >= JwtConstants.JweSegmentCount, then check afterwards that it equals Jws or Jwe.

Reply:
That makes sense. I've changed the logic here to make use of `string.Split()`, where the maximum number of substrings is equal to `JwConstants.MaxJwtSegmentCount + 1`. This way any string with greater than 5 segments will have `tokenParts.Length == 6`.